src/index.ts

```diff
@@ -1,4 +1,137 @@
-export default () => {
-  // eslint-disable-next-line no-console
-  console.log('https://www.youtube.com/watch?v=dQw4w9WgXcQ')
+import ResponseError from './ResponseError'
+
+export interface ApiClientConfiguration {
+  baseUrl: string
+  headers: Record<string, string>
 }
+
+export interface RequestOptions {
+  baseUrl?: string
+  headers?: Record<string, string>
+  params?: Record<string, string | number>
+  accessToken?: string
+}
+
+class ApiClient {
+  private configuration: ApiClientConfiguration = {
+    baseUrl: '',
+    headers: { 'Content-Type': 'application/json' },
+  }
+
+  private static replaceParameters(path: string, options?: RequestOptions) {
+    const parameters = path.match(/{(.*?)}/g) || []
+
+    if (!parameters.length) {
+      return path
+    }
+
+    let replacedPath = path
+    parameters.forEach((param) => {
+      const value = options?.params?.[param.slice(1, -1)]
+
+      if (!value) {
+        throw new Error(
+          `Param ${param} missing. Expected parameters are: ${parameters.join(
+            ', ',
+          )}`,
+        )
+      }
+
+      replacedPath = replacedPath.replace(param, `${value}`)
+    })
+    return replacedPath
+  }
+
+  private getPath(path: string, options?: RequestOptions) {
+    const baseUrl = options?.baseUrl || this.configuration.baseUrl
+    if (!baseUrl) {
+      throw new Error(
+        'ApiClient is not configured. Please run ApiClient.configure() or pass a custom baseUrl.',
+      )
+    }
+    const fetchPath = ApiClient.replaceParameters(path, options)
+    return [
+      baseUrl.endsWith('/') ? baseUrl.slice(0, -1) : baseUrl,
+      fetchPath.startsWith('/') ? fetchPath.slice(1) : fetchPath,
+    ].join('/')
+  }
+
+  private getHeaders(options?: RequestOptions): Record<string, string> {
+    return {
+      ...this.configuration.headers,
+      ...options?.headers,
+      ...(options?.accessToken
+        ? { Authorization: `Bearer ${options.accessToken}` }
+        : {}),
+    }
+  }
+
+  private static async returnData(response: Response) {
+    const size = Number(response.headers.get('content-length') || 0)
+    const data = size > 0 ? await response.json() : null
+    if (!response.ok) {
+      throw new ResponseError(response, data)
+    }
+    return data
+  }
+
+  configure = (configuration: Partial<ApiClientConfiguration>) => {
+    this.configuration = {
+      ...this.configuration,
+      ...configuration,
+    }
+  }
+
+  get = <ResponseType>(
+    path: string,
+    options?: RequestOptions,
+  ): Promise<ResponseType> => {
```

Comment:
I assume we cannot type the rejection?

Reply:
no.. the type parameter for Promise is only the type for .resolve, not for .reject. The type for .reject is always `any`, because the type cannot be known at compile time